src/http_tracker_connection.cpp

```diff
@@ -143,7 +144,7 @@ namespace libtorrent {
 				, escape_string({tracker_req().pid.data(), 20}).c_str()
 				// the i2p tracker seems to verify that the port is not 0,
 				// even though it ignores it otherwise
-				, tracker_req().listen_port
+				, announce_port ? announce_port : tracker_req().listen_port
```

Comment:
I think a better place to override the listen port is in `session_impl.cpp` where the tracker request object is set up and queued.

https://github.com/arvidn/libtorrent/blob/RC_2_0/src/session_impl.cpp#L1306-L1319

Then you don't need to change the HTTP and UDP tracker logic (I don't think at least)